File: plugins/capawesome-capacitor-file-compressor/android/src/main/java/io/capawesome/capacitorjs/plugins/filecompressor/FileCompressorPlugin.java
package io.capawesome.capacitorjs.plugins.filecompressor;

import android.graphics.Bitmap;
import android.net.Uri;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;

@CapacitorPlugin(name = "FileCompressor")
public class FileCompressorPlugin extends Plugin {

    public static final String TAG = "FileCompressorPlugin";

    public static final String ERROR_PATH_MISSING = "path must be provided.";
    public static final String ERROR_MIME_TYPE_NOT_SUPPORTED = "mimeType is not supported.";
    public static final String ERROR_FILE_NOT_EXIST = "File does not exist.";

    private FileCompressor implementation;

    @Override
    public void load() {
        implementation = new FileCompressor(this);
    }

    @PluginMethod
    public void compressImage(PluginCall call) {
        try {
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }
            String mimeType = call.getString("mimeType", "image/jpeg");
            if (!mimeType.equals("image/jpeg") && !mimeType.equals("image/webp")) {
                call.reject(ERROR_MIME_TYPE_NOT_SUPPORTED);
                return;
            }
            double quality = call.getDouble("quality", 0.6);

            Uri uri = implementation.getUriByPath(path);
            boolean fileExists = implementation.isFileExists(uri);
            if (!fileExists) {
                call.reject(ERROR_FILE_NOT_EXIST);
                return;
            }

            File file = implementation.compressImage(uri, mimeType, quality);

            JSObject result = new JSObject();
            result.put("path", file.getAbsolutePath());
            call.resolve(result);
        } catch (Exception exception) {
            String message = exception.getMessage();
            Log.e(TAG, message);
            call.reject(message);
        }
    }
}

File: plugins/capawesome-capacitor-file-compressor/ios/Plugin/FileCompressorPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FileCompressorPlugin)
public class FileCompressorPlugin: CAPPlugin {
    public let errorPathMissing = "path must be provided."
    public let errorMimeTypeNotSupported = "mimeType is not supported."
    public let errorFileNotExist = "File does not exist."

    private var implementation: FileCompressor?

    override public func load() {
        self.implementation = FileCompressor(plugin: self)
    }

    @objc func compressImage(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }
        let mimeType = call.getString("mimeType") ?? "image/jpeg"
        if mimeType != "image/jpeg" {
            call.reject(errorMimeTypeNotSupported)
            return
        }
        let quality = call.getFloat("quality") ?? 0.6

        guard let url = implementation?.getFileUrlByPath(path) else {
            call.reject(errorFileNotExist)
            return
        }

        implementation?.compressImage(url: url, mimeType: mimeType, quality: quality, completion: { path in
            var result = JSObject()
            result["path"] = path
            call.resolve(result)
        })
    }
}
